File: src/java1/algorithms/array/MaxProductSubArray.java
public class MaxProductSubArray {
    private static int maxProductSubArray(int[] nums) {
        int result = nums[0];
        int currentMax = 1, currentMin = 1;
        for(int n: nums) {
            int tempMaxProduct = currentMax * n;
            currentMax = Math.max(Math.max(tempMaxProduct, currentMin * n), n);
            currentMin = Math.min(Math.min(tempMaxProduct, currentMin * n), n);
            result = Math.max(currentMax, result);
        }
        return result;
    }

    public static void main(String[] args) {
        int[] numbers1 = {6, 7,-4, 5, 8, 1};
        System.out.println(maxProductSubArray(numbers1));

        int[] numbers2 = {2, 7,-4, 0, 3, 3};
        System.out.println(maxProductSubArray(numbers2));
    }
}

File: src/javascript/algorithms/array/maxProductSubArray.js
function maxProductSubArray(nums) {
    let result = nums[0];
    let currentMax = currentMin = 1;

    for(let n of nums) {
        let tempMaxProduct = currentMax * n;
        currentMax = Math.max(Math.max(tempMaxProduct, currentMin *n), n);
        currentMin = Math.min(Math.min(tempMaxProduct, currentMin *n), n);
        result = Math.max(result, currentMax);
    }
    return result;
}

let numbers1 = [6, 7,-4, 5, 8, 1];
console.log(maxProductSubArray(numbers1));
let numbers2 = [2, 7,-4, 0, 3, 3];
console.log(maxProductSubArray(numbers2));